README.md
# NodeJS Chess

This project is bare minimum for building a chess site with nodejs and socket.io.
It uses javascript chess libraries: Chessboard.js, Chess.js, and the Garbo Chess Engine.

Please visit the project site: http://chessbook.org

## Installation

Download the project files and unzip to a directory.

Install node.js and npm. Run below commands within the same directory.

    npm install express
    npm install socket.io
    
Put the piece images in client/img/alpha directory. Or you can modify client/js/oldchessbook.js file.

Run node with project.
    
    node .

## Features

* Guest can play againts computer (GarboChess.js, 800ms)
* Guest can play againts another random guest
* Guest can watch random game

## Not Supported

Below features are not and will not supported. You must implement them.
* Chess game by time
* Draw detection by repetition or by 50 moves
* Registered users
* Saving chess games
* ELO or other rank system
* Play againts diferrent levels of computer

# Working with GarboChess.js

To work with GarboChess.js I need to understand some other programmers code. One or two projects use Gary Linscott's 
main codes from boardui.js. I figured it out how to include only garbochess.js and not include boardui.js (it is a board ui, and we dont need it). I will explain my way. You need to modify for your needs.

First we created a worker:
```js
var g_garbo = new Worker("/js/garbochess.js");
```
To check messages from worker, we need an onmessage function:
```js
g_garbo.onmessage = function (e) {
   ...
}
```
And then send messages to worker:
```js
g_garbo.postMessage("position " + game.fen());
g_garbo.postMessage("search 800");
```
First line sets up a position, second line says "Solve it in 800 ms."

And I check messages from worker like this:
```js
g_garbo.onmessage = function (e) {
    move = { from: e.data[0]+e.data[1], to: e.data[2]+e.data[3] };
    if (e.data[4]) {
        move.promotion = e.data[4];
    }
    g_garbo.terminate();
    game.move(move);
    board.position(game.fen());
    socket.emit("move", {move: move, fen: game.fen()});
}
```
CAUTION: Since I modified GarboChess.js (latest "Test Harness" part) I only get the best move as message.
But original GarboChess.js send more messages if you like it analyze the position. Then you may need something like that:
```js
g_garbo.onmessage = function (e) {
    if (e.data.match("^pv") == "pv") {
    } else if (e.data.match("^sam") == "sam") {
    } else if (e.data.match("^message") == "message") {
    } else {
        move = { from: e.data[0]+e.data[1], to: e.data[2]+e.data[3] };
        if (e.data[4]) {
            move.promotion = e.data[4];
        }
        g_garbo.terminate();
        game.move(move);
        board.position(game.fen());
        socket.emit("move", {move: move, fen: game.fen()});
    }        
}
```


history.txt
Nodejs Chess (github.com/aydos/nodejs_chess)
Fahri Aydos (aydos.com|.net|.org)

2015 07 24 - v0.3
* local'de git kullanılmaya başlandı

2015 06 26 - v0.3
* ismi "nodejs_chess" olarak değiştirildi

2015 06 09 - v0.3
* küçük düzeltmeler
* oyun sırası gösteriliyor

2015 06 08 - v0.2
* github'a kondu

2015 06 04 - v0.1
* chessbook.org sitesinde yayınlandı

2015 05 29 - v0.0
* proje başlangıcı, ilk code, adı "Old ChessBook"



client/js/oldchessbook.js

// chess game variables
var socket, user, game, board, cfgplay, cfgwatch;

// jquery variables
var $username, $usercnt, $gamecnt, $gameinfo;
var $hider, $status, $askplay;
var $turnup, $turndown;
var $playerup, $playerdown;
var $message, $cancel;
var $watchgames;
var $watcherinfo;
var $aydos, $about;

// main function
$(document).ready(function() {
    socket = io.connect();
    //user = "";
    //game = new Chess();
    cfgplay = {
        draggable: true,
        position: "start",
        onDragStart: onDragStart,
        dropOffBoard: "snapback",
        onDrop: onDrop,
        onSnapEnd: onSnapEnd,
        pieceTheme: "img/alpha/{piece}.png"
    };
    cfgwatch = {
        position: "start",
        pieceTheme: "img/alpha/{piece}.png"
    };
    board = new ChessBoard("board", cfgwatch);

    $username = $("#username");
    $usercnt = $("#usercnt");
    $gamecnt = $("#gamecnt");
    $gameinfo = $("#gameinfo");
    $hider = $("#hider");
    $status = $("#status");
    $askplay = $("#askplay");
    $turnup = $("#turnup");
    $turndown = $("#turndown");
    $watchgames = $("#watchgames");
    $playerup = $("#playerup");
    $playerdown = $("#playerdown");
    $message = $("#message");
    $cancel = $("#cancel");
    $watcherinfo = $("#watcherinfo");
    $aydos = $("#aydos");
    $about = $("#about");

    $(window).bind("resize", initWindow);
    $(document).bind("keydown", keyDown);

    initWindow();
    clickFunctions();
    socketFunctions();

    askplay();
});


function socketFunctions() {

    socket.on("user", function (data) {
        user = data.user;
        $username.html(user.name);
        if (user.status == "idle") {
            askplay();
        }
    });

    socket.on("counts", function (data) {
        $usercnt.html(data.usercnt);
        $gamecnt.html(data.gamecnt);
        if (data.gamecnt == 0) {
            $watchgames.prop("disabled", true);
        } else {
            $watchgames.prop("disabled", false);
        }
    });

    socket.on("start game", function (data) {
        $hider.fadeOut("fast");
        $status.fadeOut("fast");
        user = data.user;
        game = new Chess();
        board = new ChessBoard("board", cfgplay);
        board.orientation(user.color);
        if(user.color == "white") {
            $turnup.removeClass("turnwhite").addClass("turnblack");
            $turndown.removeClass("turnblack").addClass("turnwhite");
        } else {
            $turnup.removeClass("turnblack").addClass("turnwhite");
            $turndown.removeClass("turnwhite").addClass("turnblack");
        }
        showTurn();
    });

    socket.on("watch game", function (data) {
        $hider.fadeOut("fast");
        $status.fadeOut("fast");
        game = new Chess(data.fen);
        board = new ChessBoard("board", cfgwatch);
        board.position(data.fen);
        board.orientation("white");
        $turnup.removeClass("turnwhite").addClass("turnblack");
        $turndown.removeClass("turnblack").addClass("turnwhite");
        showTurn();
    });

    socket.on("finish game", function (data) {
        user = data.user;
        showStatus(data.why, "OK");
    });

    socket.on("wait", function (data) {
        user = data.user;
        showStatus("Waiting for opponent", "Cancel");
    });

    socket.on("game info", function (data) {
        if (board.orientation() === "white") {
            $playerup.html(data.black);
            $playerdown.html(data.white);
        } else {
            $playerup.html(data.white);
            $playerdown.html(data.black);
        }
        if (data.watchers) {
            $watcherinfo.html("Watcher this game:<br /><span id='watchercnt'>"+data.watchers + "</span>");
        } else {
            $watcherinfo.html("");
        }
    });

    socket.on("move", function (data) {
        game.move(data.move);
        board.position(game.fen());
        showTurn();
    });

}

function initWindow() {
    var w = document.body.clientWidth;
    var h = document.body.clientHeight;
    
    var gh = h - $("#menu").outerHeight() - 24;
    var gw = gh + 400;
    $("#game").css({width: gw, height: gh});
    $("#board").css({width: gh, height: gh});
    $("#infoup").css({height: gh});
    $hider.css({width: 2*w, height: 2*h, "margin-left": (0-w), "margin-top": (0-h)});

    board.resize();
}

function keyDown(e) {
    if (e.keyCode == 27) {      // esc
        e.preventDefault();
        if (user.status === "play") {
            if ($status.is(':visible')) { // cancel resign
                $hider.fadeOut("fast");
                $status.fadeOut("fast");
            } else {                           // ask resign
                showStatus("Do you resign the game?", "Yes");
            }
        } else if (user.status === "watch") {  // leave watching game
            socket.emit("cancel watch");
        }
    }
}

function showStatus(message, button) {
    $hider.fadeIn("fast");
    $status.fadeIn("fast");
    $message.html(message);
    $cancel.html(button);
}

function showTurn() {
    var turn = game.turn() === "w" ? "white" : "black";
    if (turn === board.orientation()) {
        $turnup.hide();
        $turndown.show();
    } else {
        $turnup.show();
        $turndown.hide();
    }
}

function clickFunctions() {

    $("#btnabout").click(function() {
        $hider.fadeIn("fast");
        $about.fadeIn("fast");
    });

    $("#btnaydos").click(function() {
        $hider.fadeIn("fast");
        $aydos.fadeIn("fast");
    });

    $("#closeabout").click(function() {
        $hider.fadeOut("fast");
        $about.fadeOut("fast");
    });

    $("#closeaydos").click(function() {
        $hider.fadeOut("fast");
        $aydos.fadeOut("fast");
    });

    askplay = function() {
        $hider.fadeIn("fast");
        $askplay.fadeIn("fast");
    };

    $("#playhuman").click(function() {
        $hider.fadeOut("fast");
        $askplay.fadeOut("fast");
        socket.emit("play human");
    });

    $("#playcomputer").click(function() {
        $hider.fadeOut("fast");
        $askplay.fadeOut("fast");
        socket.emit("play computer");
    });

    $("#watchgames").click(function() {
        $hider.fadeOut("fast");
        $askplay.fadeOut("fast");
        socket.emit("watch game");
    });

    $cancel.click(function() {
        $status.fadeOut("fast");
        $askplay.fadeIn("fast");
        if ($cancel.html() === "Cancel") {
            socket.emit("cancel wait");
        }
        if ($cancel.html() === "OK") {
            ;
        }
        if ($cancel.html() === "Yes") {
            var status = {
                draw : 0,
                white : 0,
                black : 0,
                resigned : user.name,
                disconnected : ""
            };
            socket.emit("finish game", status);
        }
    });

}


// ====================================================
// Integration between chessboard.js and chess.js

// do not pick up pieces if the game is over
// only pick up pieces for the side to move
var onDragStart = function(source, piece, position, orientation) {
    if (game.game_over() === true || (game.turn() != piece[0]) ) {
        return false;
    }
    if (game.turn() === "w" && user.color === "black") {
        return false;
    }
    if (game.turn() === "b" && user.color === "white") {
        return false;
    }
};

var onDrop = function(source, target) {
    // see if the move is legal
    var move = game.move({
        from: source,
        to: target,
        promotion: "q"
    });
    // illegal move
    if (move === null) return "snapback";

    // make the move
    socket.emit("move", {move: move.san, fen: game.fen()})

    // if play againts computer
    if (user.opponent === "computer") {
        makeComputerMove();
    }

    updateStatus();
};

// update the board position after the piece snap 
// for castling, en passant, pawn promotion
var onSnapEnd = function() {
    board.position(game.fen());
    showTurn();
};

var updateStatus = function() {
    var status = {
        draw : 0,
        white : 0,
        black : 0,
        resigned : "",
        disconnected : ""
    };
    /*var moveColor = 'White';
    if (game.turn() === 'b') {
        moveColor = 'Black';
    }*/
    // checkmate?
    if (game.in_checkmate() === true) {
        if (game.turn() === 'w') {
            status.black = 1;
        } else {
            status.white = 1;
        }
        //status = 'Game over, ' + moveColor + ' is in checkmate.';
        socket.emit("finish game", status);
    }
    // draw?
    else if (game.in_draw() === true) {
        status.draw = 1;
        socket.emit("finish game", status);
    }
    // game still on
    else {
        /*status = moveColor + ' to move';
        // check?
        if (game.in_check() === true) {
            status += ', ' + moveColor + ' is in check';
        }*/
    }
};

var makeComputerMove = function() {
    var possibleMoves = game.moves();

    // game over
    if (possibleMoves.length === 0) return;

    /*
    // Random Move
    var randomIndex = Math.floor(Math.random() * possibleMoves.length);
    game.move(possibleMoves[randomIndex]);
    board.position(game.fen());
    socket.emit("move", {move: possibleMoves[randomIndex], fen: game.fen()})
    */

    // GarboChess
    var g_garbo = new Worker("/js/garbochess.min.js");
    g_garbo.onmessage = function (e) {
        move = { from: e.data[0]+e.data[1], to: e.data[2]+e.data[3] };
        if (e.data[4]) {
            move.promotion = e.data[4];
        }
        g_garbo.terminate();
        game.move(move);
        board.position(game.fen());
        showTurn();
        socket.emit("move", {move: move, fen: game.fen()});
        updateStatus();
    }
    g_garbo.postMessage("position " + game.fen());
    g_garbo.postMessage("search 800");

};



server.js
/*

OLD CHESS BOOK (http://chessbook.org)
Fahri Aydos (http://aydos.com)

v0.0 2015 05 29
- ilk code

v0.1 2015 06 04
- sitede yayınlandı

v0.2 2015 06 08
- github'a kondu

v0.3 2015 06 09
- küçük düzeltmeler
- oyun sırası

*/

var express = require('express');
var app = express();
var http = require('http').Server(app);
var io = require('socket.io').listen(http);

http.listen(8081);

app.use(express.static(__dirname + '/client'));

var wait = "";
var users = {};
var usercnt = 0;
var games = {};
var gamecnt = 0;


io.on("connection", function (socket) {

    // generate new user add to users list
    var user = newUser(socket);

    // save name for socket session
    socket.name = user.name;

    // send back user data and broadcats counters
    socket.emit("user", {user: users[socket.name]});
    broadcastCounts();

    // ==========================================
    // socket functions

    // user disconnect
    socket.on("disconnect", function () {
        deleteUser(socket.name);
        broadcastCounts();
    });

    // user want to play againts computer
    socket.on("play computer", function () {
        var id = newGame(socket.name, "computer");
        broadcastCounts();
        broadcastGameInfo(id);
    });

    // user want to play againts human
    socket.on("play human", function () {
        if (wait === "") {
            wait = socket.name;
            users[socket.name].status = "wait";
            socket.emit("wait", {user: users[socket.name]});
        } else {
            var opponent = wait;
            wait = "";
            var id = newGame(opponent, socket.name);
            broadcastCounts();
            broadcastGameInfo(id);
        }
    });

    // user dont want to wait human opponent
    socket.on("cancel wait", function () {
        if (wait === socket.name) {
            wait = "";
        }
        users[socket.name].status = "idle";
        socket.emit("user", {user: users[socket.name]});
    });

    socket.on("watch game", function () {
        var keys = Object.keys(games);
        if (keys.length > 0) {
            var id = keys[Rand(0, keys.length-1)];
            games[id].watchers.push(socket.name);
            users[socket.name].status = "watch";
            users[socket.name].game = id;
            socket.emit("user", {user: users[socket.name]});
            socket.emit("watch game", {fen : games[id].fen});
            broadcastGameInfo(id);
        }
    });

    socket.on("cancel watch", function () {
        var id = users[socket.name].game;
        if (games[id]) {
            var index = games[id].watchers.indexOf(socket.name);
            if (index > -1) {
                games[id].watchers.splice(index, 1);
            }
            broadcastGameInfo(id);
        }
        users[socket.name].status = "idle";
        users[socket.name].game = "";
        socket.emit("user", {user: users[socket.name]});
    });

    socket.on("finish game", function (data) {
        id = users[socket.name].game;
        finishGame(id, data);
        broadcastCounts();
    });

    socket.on("move", function (data) {
        var opponent = users[socket.name].opponent;
        if (opponent != "computer") {
            io.to(users[opponent].id).emit("move", {move: data.move});
        }
        var id = users[socket.name].game;
        games[id].fen = data.fen;
        for (var i=0; i<games[id].watchers.length; i++) {
            io.to(users[games[id].watchers[i]].id).emit("move", {move: data.move});
        };
    });

});

function broadcastCounts() {
    io.sockets.emit("counts", {usercnt: usercnt, gamecnt: gamecnt});
}


function broadcastGameInfo(id) {
    var info = {
        white : games[id].white,
        black : games[id].black,
        watchers : games[id].watchers.length
    }
    // send to white
    io.to(users[games[id].white].id).emit("game info", info);
    // send to black
    if (games[id].black != "computer") {
        io.to(users[games[id].black].id).emit("game info", info);
    }
    // send to watchers
    for (var i=0; i<games[id].watchers.length; i++) {
        if (users[games[id].watchers[i]]) {
            io.to(users[games[id].watchers[i]].id).emit("game info", info);
        }
    }
}

// generate new user add to users list
function newUser(socket) {
    // generate non exists name
    var name = hexRand();
    while (users[name]) {
        name = hexRand();
    }
    // new user
    var user = {
        name : name,
        status : "idle", // idle, wait, play, watch
        game : "",
        color : "",
        opponent : "",
        id : socket.id
    }
    // add to users list
    users[user.name] = user;
    usercnt++;
    return user;
}


function deleteUser(name) {
    var id = users[name].game;
    var status = users[name].status
    if (status === "play") {
        finishGame(id, name + " disconnected", name);
    } else if (status === "watch") {
        if (games[id] != undefined) {
            var index = games[id].watchers.indexOf(name);
            if (index > -1) {
                games[id].watchers.splice(index, 1);
            }
            broadcastGameInfo(id);
        }
    }
    if (wait === name) {
        wait = "";
    }
    delete(users[name]);
    usercnt--;
}


function cancelWatch() {
    var id = users[socket.name].game;
    if (games[id] != undefined) {
        var index = games[id].watchers.indexOf(name);
        if (index > -1) {
            games[id].watchers.splice(index, 1);
        }
        broadcastGameInfo(id);
    }
    users[socket.name].status = "idle";
    users[socket.name].game = "";
    users[socket.name].color = "";
    users[socket.name].opponent = "";
}


function newGame(white, black) {
    var id = hexRand();
    while (games[id] != undefined) {
        id = hexRand();
    }
    var game = {
        id : id,
        white : white,
        black : black,
        fen : "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
        watchers : []
    }
    games[id] = game;
    gamecnt++;
    users[white].status = "play";
    users[white].game = game.id;
    users[white].color = "white";
    users[white].opponent = black;
    io.to(users[white].id).emit("start game", {user: users[white]});
    if (black != "computer") {
        users[black].status = "play";
        users[black].game = game.id;
        users[black].color = "black";
        users[black].opponent = white;
        io.to(users[black].id).emit("start game", {user: users[black]});
    }
    return id;
}


function finishGame(id, data) {
    if (games[id] === undefined) {
        return;
    }
    // save users info
    var white = games[id].white;
    var black = games[id].black;
    var watchers = games[id].watchers;

    // delet game
    delete(games[id]);
    gamecnt--;

    var why = "";
    if (data.draw == 1) {
        why = "Draw";
    } else if (data.white == 1) {
        why = "White ("+ white +") won.";
    } else if (data.black == 1) {
        why = "Black ("+ black +") won.";
    } else if (data.resigned === white) {
        why = "Black ("+ black +") won. " + white + " resigned.";
    } else if (data.resigned === black) {
        why = "White ("+ white +") won. " + black + " resigned.";
    } else if (data.disconnected === white) {
        why = "Black ("+ black +") won. " + white + " disconnected.";
    } else if (data.disconnected === black) {
        why = "White ("+ white +") won. " + black + " disconnected.";
    } else {
        why = "Game finished";
    }

    // do user staff
    if (white != data.disconnected) {
        users[white].status = "idle";
        users[white].game = "";
        users[white].color = "";
        users[white].opponent = "";
        io.to(users[white].id).emit("finish game", {user: users[white], why: why});
    }
    if (black != "computer" && black != data.disconnected) { // if black not disconnected
        users[black].status = "idle";
        users[black].game = "";
        users[black].color = "";
        users[black].opponent = "";
        io.to(users[black].id).emit("finish game", {user: users[black], why: why});
    }
    watchers.forEach( function(name) {
        if(users[name] != undefined) {
            users[name].status = "idle";
            users[name].game = "";
            users[name].color = "";
            users[name].opponent = "";
            io.to(users[name].id).emit("finish game", {user: users[name], why: why});
        }
    });
}

// Random utilities

function Rand(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

function hexRand() {
    rand = "";
    chrs = "0123456789ABCDEF"
    for (var i=0; i<8; i++) {
        rand = rand + chrs[Rand(0,15)];
    };
    return rand;
}

